### ISSUE TITLE
Repository.save discussion

### ISSUE THREAD

**Issue by ivanjunckes** [JNOSQL/artemis#20]
Hello guys, I was using artemis and saw that repository is using save() method to insert entities.

I was wondering that save can be either a insertion or an update. So I think would be more appropriate to use:
repository.insert for insertions
repository.update for updates

What do you think?

**Comment by otaviojava** [comment]
That's a good point discussion.
Right now, the best NoSQL mapper reference is Spring Data and they're using the **save**

- http://docs.spring.io/spring-data/cassandra/docs/2.0.0.M2/reference/html/#repositories.core-concepts

**Comment by olivergierke** [comment]
tl;dr — according to DDD, a repository is basically an abstraction of a collection of aggregates, which you can add aggregates to, look them up (either by id, all or in a filtered way). Thus, in general, it must not (should not) expose store or persistence specific operations.

I'd als raise the question, why it should be the repository client's responsibility to distinguish between the two operations. It imposes a state change on an aggregate (which might actually be its creation) and asks the repository to persist this state change, that's it.

I gave a more detailed answer in this [StackOverflow question](https://stackoverflow.com/questions/8625150/why-to-use-returned-instance-after-save-on-spring-data-jpa-repository/8634586#8634586).

**Comment by ivanjunckes** [comment]
Makes sense @olivergierke, then I think we need to remove "update" from the repository and just keep save for insert and update. Currently we have save and update which makes a little confusing.

What do you think?

**Comment by otaviojava** [comment]
A proper nomenclature could be renamed:

- select 
- update
- insert
- delete


WDYT?

**Comment by OndrejM** [comment]
I agree with @olivergierke, the abstraction should only provide `save` method, not update/insert. The repository should have enough data to find out if the entity exists in the store and update it, or if it doesn't and insert it. If there's no

Therefore the following methods make sense:
 - select / find
 - save (instead of update/insert)
 - delete

Since the API doesn't put any requirements on the entity, it's not clear whether it contains ID or some other metadata to find a matching record in the DB to updated it. In this case, I would define that if the repository can update an entity, a call to `save` means update. If it can't update the entity, it would mean insert.

**Comment by olivergierke** [comment]
One more thought: we're [going to ship](https://jira.spring.io/browse/DATACMNS-944) a significantly renamed set of `CrudRepository` method names in Spring Data 2.0 as the previous ones have proven to cause problems when it comes to the proxy implementation and the detection which method actually needs to be called on the target repository instance. That has mostly been due to generics resolution and method overloads ending up with the same signature in corner cases. Not sure that this is an issue to you guys already but we now move to a naming scheme where method name + raw parameter type uniquely identify CRUD methods.

I can go into details if you want / need more information about this, but thought I just leave you with the gist of it for now.

**Comment by otaviojava** [comment]
Ok, based either Spring Data 2.0 and this suggestion, we can create an ID annotation and refactor the API to something like that:

```
public interface Repository<T, ID> {

    T save(T entity) throws NullPointerException;

    T save(T entity, Duration ttl);

 
    Iterable<T> save(Iterable<T> entities) throws NullPointerException;


    Iterable<T> save(Iterable<T> entities, Duration ttl) throws NullPointerException;


    void deleteById(ID id) throws NullPointerException;


    void delete(Iterable<T> entities) throws NullPointerException;

    void delete(T entity) throws NullPointerException;

    Optional<T> findById(ID id) throws NullPointerException;

    Iterable<T> findById(Iterable<ID> ids) throws NullPointerException;
  
    boolean existsById(ID id) throws NullPointerException;


}
```

**Comment by olivergierke** [comment]
Just a final comment. The reason Spring Data moved away from `delete(Iterable)` and `delete(T)` is that it might create ambiguities in bridge methods if `T` implements `Iterable` as well. The reason we need to be so picky about this is because we allow users to define a plain interface that only has matching method signatures and doesn't even extent the equivalent to your `Repository` in the first place.

A second thing: I'd change the `save(…)` method signatures to `<S extends T> S save(S entity)` as this allows callers to get the same type returned if they hand in a subclass of `T`. Imagine a class `Contact` with subtypes `Person` and `Company`. With your signature, I'd get a `Contact` returned even if I hand in a `Person`.

Final note. I don't think exposing overloads with a `Duration` is a good idea on the repository level as it's basically leaky abstraction as a repository client now has to be aware of TTLs. I'd argue, that's something that should be hidden by the repository, not exposed.

**Comment by otaviojava** [comment]
Thanks @olivergierke updated
